cityofaustin/joplin
# Issue 4356: publish queue notification

Issue: https://github.com/cityofaustin/techstack/issues/4356
Related to: https://github.com/cityofaustin/publisher/pull/13/files

This ended up being a big one. The amount of code changed across 2 repositories is overwhelming to look at without context. So here's a play-by-play walkthrough of what's been updated and why. Hopefully this will make code review easier to do.

- [ ] Added 4 new publish_* fields to `joplin/pages/base_page/models.py` to handle the real "Publishing" state. Things are no longer just "Live" or "Draft". https://github.com/cityofaustin/joplin/blob/9be665c63d0f7f79649f666bdfb08673bf7258d4/joplin/pages/base_page/models.py#L51
- [ ] The logic to set the new page_status is in the `status_string` method in `joplin/pages/base_page/models.py` https://github.com/cityofaustin/joplin/blob/9be665c63d0f7f79649f666bdfb08673bf7258d4/joplin/pages/base_page/models.py#L162
- [ ] The template to render those new status_string options is in `joplin/templates/wagtailadmin/shared/page_status_tag.html` https://github.com/cityofaustin/joplin/blob/4356-queue/joplin/templates/wagtailadmin/shared/page_status_tag.html
- [ ] We had to basically completely redo how we're collecting pages to send to the Publisher. We're now sending a lot more data than just the page_id. https://github.com/cityofaustin/joplin/blob/4356-queue/joplin/base/signals/janis_build_triggers.py
- [ ] That data gets sent to the publisher. https://github.com/cityofaustin/publisher/pull/13/files
- [ ] The page data is received by the publisher's newly updated publish_request_handler. https://github.com/cityofaustin/publisher/blob/4356-queue/src/handlers/publish_request_handler.py
- [ ] The publish_request_handler now has extensive tests to handle edge cases and errors. https://github.com/cityofaustin/publisher/blob/4356-queue/src/tests/test_publish_request_handler.py
- [ ] While our code is building, we now display fancy badges for our "Publishing" and "Unpublishing" page statuses. And if you click on them, a nice informational modal pops up! And that was really the goal for this issue. Everything else in these 2 PRs are just means to an end of displaying this 15 line html file. What is life. https://github.com/cityofaustin/joplin/blob/4356-queue/joplin/templates/joplin_UI/pageStatusModals/publishing.html
- [ ] At the end of a successful build in Publisher, it will send a response back to Joplin. https://github.com/cityofaustin/publisher/blob/4356-queue/src/handlers/commands/send_publish_succeeded_message.py
- [ ] Joplin receives this message. We've installed the django-rest-framework and an api-key library to authenticate this endpoint. That way, only the Publisher (with a valid API key) will be able to access this open endpoint. https://github.com/cityofaustin/joplin/blob/4356-queue/joplin/base/views/publish_succeeded.py

Bonus things:
- [ ] Our PR Heroku postgres databases will now use version 12, just like staging and prod.
- [ ] I added uglify compression to our production webpack builds for Joplin. It appears to have had no impact. To be continued.
- [ ] new_page_from_modal now uses a `@require_POST` decorator instead of an "if request.POST" statement inside of it. That's why it looks like there are a bunch of changes. That code basically hasn't changed, it's just unindented.
- [ ] all fixtures now have revisions, and "live" fixtures are also "published"

TODO:
- [ ] Need to ask design where this little banner goes: https://app.abstract.com/projects/bbe34f6b-7080-4dcd-9fa9-d8e30d8f6128/branches/master/commits/f90d313fba24e354093e1baffe92abaceac1fd79/files/a92ebeb2-03b6-44cb-8376-a1b7230da8bd/layers/9073443f-4254-4cf8-a52b-11e8ccbf49b8?collectionId=acbfaded-c24c-4ea0-b8de-b87e4758f854&collectionLayerId=8d4c885d-16de-4750-85f9-8a05b946d520&mode=build

## How to Test
Go to https://joplin-pr-4356-queue.herokuapp.com/admin/pages/search/. Feel free to go crazy. Publish stuff, unpublish stuff. You should see the correct page_status badges appear in accordance with the rules set forth here: https://github.com/cityofaustin/joplin/blob/9be665c63d0f7f79649f666bdfb08673bf7258d4/joplin/pages/base_page/models.py#L162

We built a distinct Publisher environment that has the new changes from the Publisher PR for this issue. You can track your builds by going to this dynamodb table: https://console.aws.amazon.com/dynamodb/home?region=us-east-1#tables:selected=coa_publisher_4356-queue;tab=items
